### ISSUE TITLE
Investigate benefits of adding BatchNorm folding to NNCF quantization

### ISSUE THREAD

**Issue by vshampor** [openvinotoolkit/nncf#88]
Original paper: https://arxiv.org/pdf/1806.08342.pdf
PyTorch implementation: https://github.com/pytorch/pytorch/blob/master/torch/nn/intrinsic/qat/modules/conv_fused.py#L82-L92

Experiment with low-batch training scenarios such as Mask-R-CNN to determine whether adding BatchNorm folding to NNCF will improve general quantized model quality.

**Comment by RikAllen** [comment]
This would I think complement the FPGA implementation very well - allowing us to remove some hardware blocks.

**Comment by ljaljushkin** [comment]
Probably, it makes sense to consider similar insight from [HAWQ-v3 paper](https://arxiv.org/pdf/2011.10680.pdf):
"We follow a simpler approach where we first keep the Conv and BN layer unfolded, and then we perform standard QAT by backpropagating the gradients as usual. After several epochs, we freeze the running statistics in the BN layer and only then perform CONV+BN folding and quantized to the target bit-width"

**Comment by vshampor** [comment]
@asenina please post current results.

**Comment by asenina** [comment]
Model | Pytorch measured | Pytorch reference FP32| Openvino AC measured
------- | -------------------- | ------------------------- | ------------------------- 
unet_camvid_int8 (with folding conv and bn) | 71. 57 % | 71.95 % | 71.67 %

**Comment by Zehaos** [comment]
@asenina  Thanks you for your work. I have several questions.

1. What is "pytorch measured", "Openvino AC measured" means? 

2. It seems that they got similar result on segmentation task. Is there a experiment on detection task?

If it is possible, I would be glad to test on more tasks using NNCF. Can the code (qat with folding conv and bn) be accessable to me?

**Comment by asenina** [comment]
@Zehaos 
1. "pytorch measured" - accuracy was measured on final compressed model in NNCF. (Use option `--mode test` and `--resume  
path-to-result-checkpoint`)
   "Openvino AC measured" - Final compressed model was converted to onnx, then to IR format. Then accuracy was measured on the model in IR format (use openvino runtime and accuraсy checker)
2. I have not done experiments on detection task.

Please, track this in  #446 . I will be glad if you test it on other task using NNCF and  share the results (if possible). Thanks

**Comment by Zehaos** [comment]
@asenina Thanks for your reply! I will go into it soon.

**Comment by asenina** [comment]
Folding was implemented according to the following formulas: 
![image](https://user-images.githubusercontent.com/66373104/111109359-a7e91f80-856b-11eb-8113-371d92a320ee.png)
[Initial implementation](https://github.com/asenina/nncf_pytorch/tree/folding_bn_with_fix)
 Results:
 Model name|	Acc (without folding) | Acc (with folding)
------- | -------------------- | ----------------|
squeezenet1_1_imagenet_int8|	58.04 %	|58.14 %
unet_camvid_int8	|71.66 %	|72.03 %
icnet_camvid_int8	|67.78 %	|67.45 %
mobilenet_v2_imagnet_int8	|71.31 %	|71.67 %
mobilenet_v2_imagnet_int4_int8	|71.03 %	|70.98 %


BatchNorm folding gives minor improvements of accuracy only for int8 models. It will not merge to NNCF.
